flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleaner.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.dispatcher.cleanup;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.util.concurrent.FutureUtils;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+/** {@code DefaultResourceCleaner} is the default implementation of {@link ResourceCleaner}. */
+public class DefaultResourceCleaner implements ResourceCleaner {
+
+    private final Collection<BiFunction<JobID, Executor, CompletableFuture<Void>>>
+            prioritizedOrderedJobRelatedCleanups = new ArrayList<>();
+    private final Collection<BiFunction<JobID, Executor, CompletableFuture<Void>>>
+            jobRelatedCleanups = new ArrayList<>();
+    private final Executor cleanupExecutor;
+
+    DefaultResourceCleaner(Executor cleanupExecutor) {
```

Comment:
It would be good to also inject the `ComponentMainThreadExecutor` here and assert that cleanupAsync is executed in the main thread as some implementation rely on it (eg. JobManagerRunnerRegistry)

Reply:
I made the components thread-safe 👍